### PR TITLE
Fix cake day calculation and display string

### LINKED ISSUE #2781: Cake day calculation is incorrect
Apparently it shows when an account is 11 months old

### PR DESCRIPTION
**Pull request by taiheng**
## Issues

- closes #2781

## Description

Anniversary dates could appear to be a day early since date strings from API assume UTC. 
This fixes the issue where anniversary date display string and logic could sometimes display an anniversary the day before in certain time zones.

## Implementation Notes

This logic appears to be exclusively used in the cake day date string and logic for calculation limiting scope of this change.

### CODE REVIEW

File: MlemTests/Extensions/DateTests.swift [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+    @Test func isDateDayOfAnniversaryIsTrue() {
+        let calendar = Calendar.current
+        let anniversaryDate = Date(timeIntervalSince1970: 771470149.5109999)
+        let testDate = calendar.date(from: DateComponents(year: 2026, month: 6, day: 13))!
+
+        #expect(anniversaryDate.isAnniversaryDate(testDate) == true)
+    }
+
+    @Test func isLeapYearDateAnniversaryInFebIsTrue() {
```
Comment (taiheng) [inline MlemTests/Extensions/DateTests.swift]: Another edge case exists for anniversary dates that fall on leap years and is currently unhandled.

File: Mlem/App/Utility/Extensions/Date+Extensions.swift [hunk 2]
```diff
@@ -55,12 +56,23 @@ extension Date {
     }
     
     var isAnniversaryToday: Bool {
-        let calendar = Calendar.current
+        return isAnniversaryDate(.now)
+    }
+
+    func isAnniversaryDate(_ otherDate: Date) -> Bool {
+        var calendar = Calendar.current
+        calendar.timeZone = TimeZone(identifier: "UTC")!
```
Comment (Sjmarf) [inline Mlem/App/Utility/Extensions/Date+Extensions.swift]: https://developer.apple.com/documentation/foundation/timezone/gmt

```suggestion
        calendar.timeZone = TimeZone.gmt
```

File: MlemTests/Extensions/DateTests.swift [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+    @Test func isDateAtOneYearAnniversaryIsTrue() {
+        let calendar = Calendar.current
+        let anniversaryDate = calendar.date(from: DateComponents(year: 2024, month: 1, day: 1))!
+        let testDate = calendar.date(from: DateComponents(year: 2025, month: 1, day: 1))!
+
+        #expect(anniversaryDate.isAnniversaryDate(testDate) == true)
+    }
+
+    @Test func isDateDayBeforeAnniversaryIsFalse() {
```
Comment (taiheng) [inline MlemTests/Extensions/DateTests.swift:45]: This date uses the parsed date from `nocturne`'s user account

File: Mlem/App/Utility/Extensions/Date+Extensions.swift
Comment (taiheng) [inline Mlem/App/Utility/Extensions/Date+Extensions.swift:29]: As far as I could tell this is only used in `ProfileDateView`

File: Mlem/App/Utility/Extensions/Date+Extensions.swift [hunk 4]
```diff
@@ -31,6 +31,7 @@ extension Date {
         dateFormatter.dateStyle = .short
         dateFormatter.timeStyle = .none
         dateFormatter.locale = Locale.current
+        dateFormatter.timeZone = TimeZone(identifier: "UTC")!
```
Comment (Sjmarf) [inline Mlem/App/Utility/Extensions/Date+Extensions.swift]: https://developer.apple.com/documentation/foundation/timezone/gmt

```suggestion
        calendar.timeZone = TimeZone.gmt
```

File: MlemTests/Extensions/DateTests.swift (hunk 1 above)
Comment (Sjmarf) [reply MlemTests/Extensions/DateTests.swift]: Good spot. IMO showing the cake day on February 28th is the better choice here. 

The downside is that it won't quite be 365 days from their account creation date to February 28th, but I think it's more intuitive to keep it within February.

**Review by Sjmarf — COMMENTED** [review]
Nice work! Very rigorous 😄

File: Mlem/App/Utility/Extensions/Date+Extensions.swift (hunk 4 above)
Comment (taiheng) [reply Mlem/App/Utility/Extensions/Date+Extensions.swift]: I forgot about that, I'll make the change.

File: MlemTests/Extensions/DateTests.swift (hunk 1 above)
Comment (taiheng) [reply MlemTests/Extensions/DateTests.swift]: @Sjmarf  Do you want this in this PR or create a separate one for leap year?

Comment (Sjmarf) [reply MlemTests/Extensions/DateTests.swift]: In this PR would be great 🙂

File: Mlem/App/Utility/Extensions/Date+Extensions.swift [hunk 5]
```diff
@@ -55,12 +56,34 @@ extension Date {
...
-        return date.month == current.month && date.day == current.day && date.year != current.year
+        let current = calendar.dateComponents([.month, .day, .year], from: otherDate)
+
+        // Year component must be explicitly unwrapped for less than comparison
+        guard let dateYear = date.year, let currentYear = current.year else { return false }
+
+        // If our target day occurred in a leap year, adjust to Feb 28 if it
+        // is being compared to a non leap year
+        let targetDay: Int?
+        if date.month == 2,
```
Comment (taiheng) [inline Mlem/App/Utility/Extensions/Date+Extensions.swift:74]: Here be magic numbers, I'm not aware how you guys handle them. I can move into a constant or something if needed.

File: MlemTests/Extensions/DateTests.swift (hunk 1 above)
Comment (taiheng) [reply MlemTests/Extensions/DateTests.swift]: Done

File: Mlem/App/Utility/Extensions/Date+Extensions.swift (hunk 5 above)
Comment (Sjmarf) [reply Mlem/App/Utility/Extensions/Date+Extensions.swift:74]: I think it's fine to leave this as-is, as we won't need to change these constants later.

**Review by Sjmarf — APPROVED** [review]
Nice! LGTM 💯
